### ISSUE TITLE
version 0.24 vulkano::shader! macro fails to recognize storage buffers

### ISSUE THREAD

**Issue by apessino** [vulkano-rs/vulkano#1627]
Hello!

In the process of updating a project from 0.23 to 0.24 I noticed that the shader! macro now fails to recognize all 9 of the storage buffers defined in my shader. The shader defines several storage buffers similar to this:

```
struct Light {
    vec3 def; // light definiton: pos if radius > 0.0, dir if radius == 0.0
    float radius; // light radius: if 0.0 the light is directional and def is its direction

    vec3 color; // light color
    float intensity; // normalized light power

    float range; // distance, in meters, from full intensity to complete light falloff
    float falloff; // exponent for the falloff function over the range, 0.0 = constant, 1.0 = linear, etc.

    vec2 shadow_trange; // distance range
    vec2 shadow_step; // distance per step
    uint shadow_samples; // number of steps
    float shadow_strength; // overall shadow strength
};

layout(set = 0, binding = 2) readonly buffer Lights {
    Light t[];
} lights;
```

I did a cargo expand and verified that the generated layout has the `storage` flag set to false for all of the buffers.

```
                                Some(DescriptorDesc {
                                    ty: DescriptorDescTy::Buffer(DescriptorBufferDesc {
                                        dynamic: None,
                                        storage: false,
                                    }),
                                    array_count: 1u32,
                                    stages: ShaderStages {
                                        fragment: true,
                                        ..ShaderStages::none()
                                    },
                                    readonly: true,
                                }),
```

I am not sure what could have changed since this has worked fine for a long time until 0.24 - I tried removing the `readonly` attribute to see if that was the problem, but the generated layout seems to be the same either way.

Any idea of what is going on?

Thanks in advance!

A=

**Comment by Rua** [comment]
I've investigated this a bit, and it looks like for some reason, the shader compiler itself is generating different code. For the basic-compute-shader example, it generates `StorageClass` 12 in 0.23 (which stands for storage buffer), but 2 (which stands for uniform) in 0.24. I'll keep looking deeper to see why this is happening.

**Comment by Rua** [comment]
I figured out the error. Storage buffers are only available from SPIR-V version 1.3 onwards, and Vulkan 0.24 changed the way the SPIR-V version is specified, so that it defaults to 1.0 unless you specify something else. So in order to use storage buffers, you need to specify either `vulkan_version: "1.1"` on the `shader!` macro, or `spirv_version: "1.3"`.

It is a bit unfortunate that the shader compiler doesn't give an error in this case. Maybe there are some compiler settings that can be changed so it's more verbose here. I'll have a look at it.

**Comment by apessino** [comment]
That was it - thanks a BUNCH, once again! 

A=
